nat/nat_status_v2.go

```diff
@@ -0,0 +1,169 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package nat
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/core/node"
+
+	"github.com/mysteriumnetwork/node/core"
+
+	"github.com/mysteriumnetwork/node/config"
+	"github.com/mysteriumnetwork/node/identity"
+)
+
+// NATStatusV2 enum
+// TODO: V2 suffix should be removed once previous tracker is removed
+type NATStatusV2 string
+
+const (
+	// Online enum
+	Online NATStatusV2 = "online"
+	// Offline enum
+	Offline NATStatusV2 = "offline"
+	// Failed enum
+	Failed NATStatusV2 = "failed"
+	// Pending enum
+	Pending NATStatusV2 = "pending"
+)
+
+// AppTopicNATStatusUpdate nat status update topic for event bus
+const AppTopicNATStatusUpdate = "AppTopicNATStatusUpdate"
+
+type currentIdentity interface {
+	GetUnlockedIdentity() (identity.Identity, bool)
+}
+
+type publisher interface {
+	Publish(topic string, data interface{})
+}
+
+// ProviderSessions should return provider session monitoring state
+type ProviderSessions func(providerID string) []Session
+
+// Session represent session monitoring state
+type Session struct {
+	ProviderID       string
+	ServiceType      string
+	MonitoringFailed bool
+}
+
+// V2NatStatusEvent nat status event
+type V2NatStatusEvent struct {
+	Status NATStatusV2
+}
+
+// StatusTrackerV2 tracks NAT status for service
+type StatusTrackerV2 struct {
+	lock   sync.Mutex
+	status NATStatusV2
+
+	publisher publisher
+
+	providerSessions ProviderSessions
+	currentIdentity  currentIdentity
+
+	pollInterval time.Duration
+}
+
+// NewStatusTrackerV2 constructor
+func NewStatusTrackerV2(
+	providerSessions ProviderSessions,
+	currentIdentity currentIdentity,
+	publisher publisher,
+	options node.OptionsNATStatusTrackerV2,
+) *StatusTrackerV2 {
+	validatedInterval := options.PollInterval
+	if validatedInterval < time.Minute {
+		validatedInterval = time.Minute
+	}
+	keeper := &StatusTrackerV2{
+		providerSessions: providerSessions,
+		currentIdentity:  currentIdentity,
+		publisher:        publisher,
+		pollInterval:     validatedInterval,
+	}
+	// consumers don't need service NAT status
+	if !config.GetBool(config.FlagConsumer) {
+		keeper.startPolling()
+	}
+	return keeper
+}
+
+func (k *StatusTrackerV2) startPolling() {
+	k.updateAndAnnounce()
+	go func() {
+		for {
+			select {
+			case <-time.After(k.pollInterval):
+				k.updateAndAnnounce()
+			}
+		}
+	}()
+}
+
+func (k *StatusTrackerV2) updateAndAnnounce() {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+	id, ok := k.currentIdentity.GetUnlockedIdentity()
+	if ok {
+		k.status = resolveNATStatus(k.providerSessions(id.Address))
+	} else {
+		k.status = Pending
+	}
+	k.publisher.Publish(AppTopicNATStatusUpdate, V2NatStatusEvent{Status: k.status})
+}
+
+// StatusForce force update status and return current result
+func (k *StatusTrackerV2) StatusForce() NATStatusV2 {
+	k.lock.Lock()
```

Comment:
This will lock mutex then call `updateAndAnnounce` which will lock the mutex again.